vowpalwabbit/automl.cc

```diff
@@ -0,0 +1,743 @@
+
+  if (!read) msg << "_aml_config_index " << config_index << "\n";
+  bin_text_read_write_fixed(
+      model_file, reinterpret_cast<char*>(&config_index), sizeof(config_index), "", read, msg, text);
+
+  chisq.save_load(model_file, read, text);
+}
+
+void scored_config::persist(metric_sink& metrics, const std::string& suffix)
+{
+  metrics.int_metrics_list.emplace_back("upcnt" + suffix, update_count);
+  metrics.float_metrics_list.emplace_back("ips" + suffix, current_ips());
+  distributionally_robust::ScoredDual sd = chisq.recompute_duals();
+  metrics.float_metrics_list.emplace_back("bound" + suffix, static_cast<float>(sd.first));
+  metrics.float_metrics_list.emplace_back("w" + suffix, last_w);
+  metrics.float_metrics_list.emplace_back("r" + suffix, last_r);
+  metrics.int_metrics_list.emplace_back("conf_idx" + suffix, config_index);
+}
+
+float scored_config::current_ips() const { return ips / update_count; }
+
+void scored_config::reset_stats()
+{
+  chisq.reset(0.05, 0.999);
+  ips = 0.0;
+  last_w = 0.0;
+  last_r = 0.0;
+  update_count = 0;
+}
+
+void exclusion_config::save_load_exclusion_config(io_buf& model_file, bool read, bool text)
+{
+  if (model_file.num_files() == 0) { return; }
+  std::stringstream msg;
+
+  size_t exclusion_size;
+  if (read)
+  {
+    bin_text_read_write_fixed(
+        model_file, reinterpret_cast<char*>(&exclusion_size), sizeof(exclusion_size), "", read, msg, text);
+    for (size_t i = 0; i < exclusion_size; ++i)
+    {
+      namespace_index ns;
+      size_t other_ns_size;
+      std::set<namespace_index> other_namespaces;
+      bin_text_read_write_fixed(model_file, (char*)&ns, sizeof(ns), "", read, msg, text);
+      bin_text_read_write_fixed(
+          model_file, reinterpret_cast<char*>(&other_ns_size), sizeof(other_ns_size), "", read, msg, text);
+      for (size_t i = 0; i < other_ns_size; ++i)
+      {
+        namespace_index other_ns;
+        bin_text_read_write_fixed(model_file, (char*)&other_ns, sizeof(other_ns), "", read, msg, text);
+        other_namespaces.insert(other_ns);
+      }
+      exclusions[ns] = other_namespaces;
+    }
+  }
+  else
+  {
+    exclusion_size = exclusions.size();
+    msg << "exclusion_size " << exclusion_size << "\n";
+    bin_text_read_write_fixed(
+        model_file, reinterpret_cast<char*>(&exclusion_size), sizeof(exclusion_size), "", read, msg, text);
+    for (const auto& ns_pair : exclusions)
+    {
+      namespace_index ns = ns_pair.first;
+      size_t other_ns_size = ns_pair.second.size();
+      bin_text_read_write_fixed(model_file, (char*)&ns, sizeof(ns), "", read, msg, text);
+      bin_text_read_write_fixed(
+          model_file, reinterpret_cast<char*>(&other_ns_size), sizeof(other_ns_size), "", read, msg, text);
+      for (const auto& other_ns : ns_pair.second)
+      { bin_text_read_write_fixed(model_file, (char*)&other_ns, sizeof(other_ns), "", read, msg, text); }
+    }
+  }
+
+  if (!read || true)
```

Comment:
This will always evaluate to true

Reply:
That was a placeholder for the versions check. Thats now made in the main automl save_load so this `if` can be removed